### PR TITLE
Feat/crystal

### PR DESCRIPTION
**Pull request by area363**
**New Added Tables**
* `Grindings`
* `UnlockEquipmentRecipes`
* `UnlockWorlds`
* `ReplaceCombinationEquipmentMaterials`
* `HasRandomBuffs`
* `HasWithRandomBuffs`
* `JoinArenas`
* `BattleArenas`

### CODE REVIEW

File: NineChronicles.DataProvider/RenderSubscriber.cs [hunk 1]
```diff
@@ -355,6 +436,69 @@ protected override Task ExecuteAsync(CancellationToken stoppingToken)
                                     SubRecipeId = combinationEquipment.subRecipeId ?? 0,
                                     BlockIndex = ev.BlockIndex,
                                 });
+                                if (combinationEquipment.payByCrystal)
+                                {
+                                    Currency crystalCurrency = new Currency("CRYSTAL", 18, minters: null);
```
Comment (ipdae) [inline NineChronicles.DataProvider/RenderSubscriber.cs:441]:
```suggestion
                                    Currency crystalCurrency = CrystalCalculator.CRYSTAL
```

File: NineChronicles.DataProvider/RenderSubscriber.cs [hunk 2]
```diff
@@ -846,6 +1038,398 @@ protected override Task ExecuteAsync(CancellationToken stoppingToken)
                                 var end = DateTimeOffset.Now;
                                 Log.Debug("Stored MigrateMonsterCollection action in block #{index}. Time Taken: {time} ms.", ev.BlockIndex, (end - start).Milliseconds);
                             }
+
+                            if (ev.Action is Grinding grind)
+                            {
+                                _renderCount++;
```
Comment (ipdae) [inline NineChronicles.DataProvider/RenderSubscriber.cs:1044]: What is this variable used for?

Comment (area363) [reply NineChronicles.DataProvider/RenderSubscriber.cs:1044]: Tldr, the DP rendersubscriber can't catch up with the chain tip if it inserts data for each action. Therefore, I've set up a rendercount that inserts all accumulated data when the count hits 500(this limit is just randomly selected which seems to be working). In short, the DP inserts data into mysql for every 500 subscribed actions.

File: NineChronicles.DataProvider/RenderSubscriber.cs (hunk 1 above)
Comment (area363) [reply NineChronicles.DataProvider/RenderSubscriber.cs:441]: 👍 I'll change other codes that uses the crystal currency.

**Comment by area363** [comment]
The remaining bugs have been addressed during the internal cluster testing and it is now good to go. I'll merge this PR, thanks for reviewing 🙏. @ipdae

**Pull request merged**